File: monster-battle.r
# Create the monster class
Monster <- list(
  name = "",
  health = 0,
  attack = 0,
  defense = 0,
  speed = 0
)

# Create the player and enemy monsters
player <- Monster
player$name <- "Player"
player$health <- 100
player$attack <- 10
player$defense <- 5
player$speed <- 8

enemy <- Monster
enemy$name <- "Enemy"
enemy$health <- 80
enemy$attack <- 8
enemy$defense <- 3
enemy$speed <- 6

# Define the battle function
battle <- function(player, enemy) {
  # Loop until one monster has no health
  while(player$health > 0 && enemy$health > 0) {
    # Determine which monster attacks first based on speed
    if(player$speed > enemy$speed) {
      # Player attacks
      damage <- player$attack - enemy$defense
      if(damage < 0) {
        damage <- 0
      }
      enemy$health <- enemy$health - damage
      cat(player$name, "attacks", enemy$name, "for", damage, "damage.\n")
      
      # Check if enemy is defeated
      if(enemy$health <= 0) {
        cat(enemy$name, "has been defeated!\n")
        break
      }
      
      # Enemy attacks
      damage <- enemy$attack - player$defense
      if(damage < 0) {
        damage <- 0
      }
      player$health <- player$health - damage
      cat(enemy$name, "attacks", player$name, "for", damage, "damage.\n")
      
      # Check if player is defeated
      if(player$health <= 0) {
        cat(player$name, "has been defeated!\n")
        break
      }
    } else {
      # Enemy attacks
      damage <- enemy$attack - player$defense
      if(damage < 0) {
        damage <- 0
      }
      player$health <- player$health - damage
      cat(enemy$name, "attacks", player$name, "for", damage, "damage.\n")
      
      # Check if player is defeated
      if(player$health <= 0) {
        cat(player$name, "has been defeated!\n")
        break
      }
      
      # Player attacks
      damage <- player$attack - enemy$defense
      if(damage < 0) {
        damage <- 0
      }
      enemy$health <- enemy$health - damage
      cat(player$name, "attacks", enemy$name, "for", damage, "damage.\n")
      
      # Check if enemy is defeated
      if(enemy$health <= 0) {
        cat(enemy$name, "has been defeated!\n")
        break
      }
    }
  }
}

# Start the battle
cat("A wild", enemy$name, "appears!\n")
battle(player, enemy)

File: monster-battle.rb
# Monster battle game in Ruby

class Monster
  attr_accessor :name, :health, :attack
  
  def initialize(name, health, attack)
    @name = name
    @health = health
    @attack = attack
  end
  
  def attack(target)
    damage = rand(@attack..@attack+5)
    target.health -= damage
    puts "#{@name} attacked #{target.name} for #{damage} damage!"
  end
end

# Main program
puts "Welcome to Monster Battle!"

# Create two monsters
monster1 = Monster.new("Goblin", 30, 5)
monster2 = Monster.new("Dragon", 50, 10)

# Loop until one of the monsters is defeated
until monster1.health <= 0 || monster2.health <= 0
  # Monster 1 attacks Monster 2
  monster1.attack(monster2)
  break if monster2.health <= 0
  puts "#{monster2.name} health: #{monster2.health}"
  
  # Monster 2 attacks Monster 1
  monster2.attack(monster1)
  break if monster1.health <= 0
  puts "#{monster1.name} health: #{monster1.health}"
end

# Declare the winner
if monster1.health <= 0
  puts "#{monster2.name} wins!"
else
  puts "#{monster1.name} wins!"
end
